### PR TITLE
Change scaleup to set_scale - supported scale other than just 1 node

### PR DESCRIPTION
**Pull request by omertuc**
First step for fixing #129

Rename scaleup role to ensure_machineset which makes sure there's at-least one worker machineset with a given instance type.
If there's no such machineset, it'll create one out of an existing worker machineset (like it does now), but now with 0 replicas instead of 1.

Added new role, called set_scale which receives an instance type and a scale param which:

Ensures that the sum of spec.replicas of all machinesets with the given instance type is equal to the scale param.

If it's not, scales down all of them to 0 except for the first one which will have its spec.replicas set to the scale param.

In any case, it waits for all of machinesets to have:
`status.availableReplicas == spec.replicas`
and
`status.replicas == spec.replicas`

Also added `__pycache__` to `.gitignore`

### CODE REVIEW

File: roles/ensure_machineset/scripts/reconfigure_machineset.sh
Comment (omertuc) [inline roles/ensure_machineset/scripts/reconfigure_machineset.sh:17]: set_scale will later on take care of making sure this value is equal to the requested user value.

reconfigure_machineset's job now is to simply make sure a machineset with that given instance type exists, so 0-replica makes more sense

File: roles/ensure_machineset/tasks/derive_machineset.yml
Comment (kpouget) [inline roles/ensure_machineset/tasks/derive_machineset.yml:13]: could you remove this task, and use
```
> {{ artifact_extra_logs_dir }}/source_machineset.json
```
instead of 
```
{{ machineset_tempfile.path }}
```
?

(` {{ artifact_extra_logs_dir }}` used to be an optional variable, but now we're heavily relying on it to store the artifacts required to precisely understand what's going on during the toolbox script execution)

File: roles/ensure_machineset/tasks/derive_machineset.yml
Comment (kpouget) [inline roles/ensure_machineset/tasks/derive_machineset.yml:26]: same here, 

> {{ artifact_extra_logs_dir }}/new_machineset.json

File: roles/ensure_machineset/tasks/derive_machineset.yml
Comment (kpouget) [inline roles/ensure_machineset/tasks/derive_machineset.yml:20]: can you indent these lines with 3 spaces, to align with `get`

File: roles/set_scale/meta/main.yml
Comment (kpouget) [inline roles/set_scale/meta/main.yml:4]: hum, I never linked roles this way, I included them explicitly in the `main.yml`
could you add a comment at the beginning of `roles/set_scale/tasks/main.yml ` to mention that this role will be executed before `roles/set_scale/tasks/main.yml ` starts?

File: toolbox/cluster/set_scale.sh
Comment (kpouget) [inline toolbox/cluster/set_scale.sh:21]: Could you add a ` roles/set_scale/default/config.yml` file to define `machineset_instance_type` and `scale`, with a comment line explaining their purpose, and setting a default value if relevant?

and at the beginning of `roles/set_scale/tasks/main.yml`, make sure that the variables are defined / populated, eg:
```
fail: ...
when: not scale
```

it's not clear to me what you are testing here with the 3 `if`, and which parameters are mandatory, which are optional.

for the mandatory one in the `USAGE`, I use `<name>` and `[name]` for optional ones

File: build/root/var/lib/ci-artifacts/entrypoints.d/gpu_operator_run_libs.sh
Comment (kpouget) [inline build/root/var/lib/ci-artifacts/entrypoints.d/gpu_operator_run_libs.sh:16]: for testing & coverage purposes, I would prefer this line to be written as
```
toolbox/cluster/ensure_machineset.sh g4dn.xlarge
toolbox/cluster/set_scale.sh g4dn.xlarge 1
```
I saw that `set_scale.sh` calls `ensure_machineset` role as a dependency,
but by calling it explicitly, we ensure that `ensure_machineset.sh` script & playbook don't break without us noticing it!

Comment (omertuc) [reply build/root/var/lib/ci-artifacts/entrypoints.d/gpu_operator_run_libs.sh:16]: I got rid of `ensure_machineset.sh` because I felt it was a bit of an implementation detail not relevant from a toolbox user POV, so there's no longer anything that can break without us noticing.

Let me know if you disagree with that choice

File: toolbox/cluster/set_scale.sh
Comment (omertuc) [reply toolbox/cluster/set_scale.sh:21]: Everything is required. The last 2 `if`s are a silly mistake, got rid of them.

File: build/root/var/lib/ci-artifacts/entrypoints.d/gpu_operator_run_libs.sh
Comment (kpouget) [reply build/root/var/lib/ci-artifacts/entrypoints.d/gpu_operator_run_libs.sh:16]: indeed, it makes sense not to have `ensure_machineset.sh`

File: toolbox/cluster/set_scale.sh
Comment (kpouget) [reply toolbox/cluster/set_scale.sh:21]: `${2:-}` is the same as `${2-}` ? .... I need to take time to learn bash extended variable support :#

Comment (omertuc) [reply toolbox/cluster/set_scale.sh]: Not sure, just copied it from https://github.com/openshift-psap/ci-artifacts/blob/7d033e63dc650b37d7c634080e9511af4f3d0d1e/toolbox/cluster/scaleup.sh#L6

**Comment by omertuc** [comment]
\test gpu-operator-e2e

File: roles/ensure_machineset/tasks/derive_machineset.yml
Comment (kpouget) [inline roles/ensure_machineset/tasks/derive_machineset.yml:20]: could you add 
```
# The instance type to user for the machineset being scaled
machineset_instance_type: ""
```
to `roles/ensure_machineset/defaults/main/config.yml`, so that the role defines its configuration variables?

File: playbooks/set_scale.yml
Comment (kpouget) [inline playbooks/set_scale.yml:7]: can you rename the playbook & role to `cluster_set_scale`?

File: playbooks/ensure_machineset.yml
Comment (kpouget) [inline playbooks/ensure_machineset.yml:7]: can you rename the playbook and role to `cluster_ensure_machineset`?

File: playbooks/ensure_machineset.yml
Comment (kpouget) [inline playbooks/ensure_machineset.yml:2]: `Ensures that the cluster has ...`?

File: playbooks/set_scale.yml
Comment (kpouget) [inline playbooks/set_scale.yml:2]: `Ensures that the cluster has the request number of machines of a given machine-type` ?

File: roles/set_scale/defaults/main/config.yml
Comment (kpouget) [inline roles/set_scale/defaults/main/config.yml:3]: I think you can remove the default value here, as the instance type is mandatory in the toolbox script

File: roles/set_scale/tasks/main.yml
Comment (kpouget) [inline roles/set_scale/tasks/main.yml:5]: I wonder what happens when `scale = 0`, did you try?

File: roles/set_scale/tasks/main.yml
Comment (kpouget) [inline roles/set_scale/tasks/main.yml:29]: can you add new lines between tasks?

File: roles/set_scale/tasks/main.yml
Comment (kpouget) [inline roles/set_scale/tasks/main.yml:39]: I'm not sure modifying *all* the machinesets [with the given machine-type] is the best strategy,
maybe you could change it to something like:

is there only one MachineSet generating type `machine-type`? yes, modify it to `scale`
no --> cannot perform task ==> allow receiving a machineset name change change only this replica to `scale`

File: roles/set_scale/tasks/main.yml
Comment (kpouget) [inline roles/set_scale/tasks/main.yml:60]: no, not directly, the machine go live in parallel!

**Review by kpouget — CHANGES_REQUESTED** [review]

File: playbooks/set_scale.yml
Comment (omertuc) [reply playbooks/set_scale.yml:2]: I prefer to use [instance type](https://aws.amazon.com/ec2/instance-types/) rather than machine type, it's also the name of the variable (`machineset_instance_type`)

File: roles/set_scale/tasks/main.yml
Comment (omertuc) [reply roles/set_scale/tasks/main.yml:5]: Good point. Just tested - even when scale is `0`, seems like it's actually being treated as `"0"` which is a non empty string and thus `not scale` is still negative.

File: roles/set_scale/tasks/main.yml
Comment (omertuc) [reply roles/set_scale/tasks/main.yml:60]: That's what I thought, will remove the TODO because your confirmation made me more confident

File: roles/set_scale/tasks/main.yml
Comment (omertuc) [reply roles/set_scale/tasks/main.yml]: I thought the idea was to make sure the number of machines of given instance type, across the entire cluster, is exactly `scale`.

If we ask the user to provide us with a machineset name that needs to be modified - then the role doesn't really do its job, it only modifies that specific machineset, and doesn't achieve the goal of setting the cluster-wide total to exactly `scale`.

Here's an alternative suggestion - if there are other machinesets in the cluster with this instance type that need to be downscaled to 0 in order to achieve the required cluster-wide `scale`, then the role will ask the user to provide a --force flag telling them it's going to happen, so we don't perform anything extreme without explicit confirmation from the user.

File: playbooks/set_scale.yml
Comment (kpouget) [reply playbooks/set_scale.yml]: sure, `instance-type` is the correct one, machine-type was a typo :)
my suggestion was more for having a shorter description

File: roles/set_scale/tasks/main.yml
Comment (kpouget) [reply roles/set_scale/tasks/main.yml]: > I thought the idea was to make sure the number of machines of given instance type, across the entire cluster, is exactly `scale`.

The main (and only) requirement is to allow the CI to scale up and down the # of GPU nodes,
so in this environment, we know that only one MachineSet of the given instance-type will be available.
The question could be solved by aborting the play if multiple MachineSets are found.

But I like the `--force` option, it makes sense to me: default behavior is to do things as simple as possible, force behavior makes sure we have the right count, possibility by scaling down some machinesets: :+1:

File: toolbox/cluster/set_scale.sh
Comment (kpouget) [inline toolbox/cluster/set_scale.sh:8]: Could you add more information about the usage?
in particular, could you mention here some the instance-types that offer GPUs?

and I was wondering how much of it is AWS specific? I was wondering if the name should be `set_scale_aws.sh` or not

File: roles/cluster_set_scale/tasks/main.yml
Comment (kpouget) [inline roles/cluster_set_scale/tasks/main.yml:59]: there is a mis-alignment here, `"name"` is one space before the rest :)

File: roles/cluster_set_scale/tasks/main.yml
Comment (kpouget) [inline roles/cluster_set_scale/tasks/main.yml:45]: nice, this wait for the Node to be running, I wasn't sure!
Could you add a comment saying that the `.status.availableReplicas` counter tracks the # of nodes running, and not only the # of `Machines`? (I though it would be the latest only)

File: toolbox/cluster/set_scale.sh
Comment (omertuc) [reply toolbox/cluster/set_scale.sh:8]: The `.sh` file itself can forever be agnostic I believe, all platforms just need an instance type name and a number to scale.

The playbooks however might already be aws-specific, not sure. Since we don't have backwards compatibility concerns, I believe it will be easy to generalize them once we want to, so no need to worry about it for now.

File: toolbox/cluster/set_scale.sh
Comment (kpouget) [inline toolbox/cluster/set_scale.sh:8]: can you mention explicitly `g4dn.xlarge` as an example of cheap GPU node, and `m4.large` as the default worker node type (correct me if I'm wrong on this one)

File: toolbox/cluster/set_scale.sh
Comment (kpouget) [inline toolbox/cluster/set_scale.sh:5]: I don't think this is the intended tests: if the 2nd and 3rd arguments are empty --> usage

maybe: `[[ -z "${1:-}" || -z "${2:-} || "${3:---force}" != "--force""`?
if 1 or 2 are empty, or 3 isn't set or is equal to `--force`

File: toolbox/cluster/set_scale.sh
Comment (kpouget) [inline toolbox/cluster/set_scale.sh:28]: you can reuse `$INSTANCE_TYPE` here, instead of `${1}`

File: README.md
Comment (kpouget) [inline README.md:165]: maybe you can add `[--force]` here, and a line of comment about it

we'll move that to the documentation soon anyway

**Review by kpouget — CHANGES_REQUESTED** [review]
thanks for the update @omertuc , I pointed out a few nits (mostly in the toolbox script)

File: toolbox/cluster/set_scale.sh
Comment (omertuc) [reply toolbox/cluster/set_scale.sh:8]: If unspecified, [the installer chooses between m4.large and m5.large according to availability](https://github.com/openshift/installer/blob/7a6374014536d5993e444e961812dd8adb2ed3fd/pkg/asset/machines/worker.go#L159)

**Comment by kpouget** [comment]
Looks good @omertuc, thanks :)

/approve

needs to rebase before LGTM

**Comment by kpouget** [comment]
Looks good @omertuc, thanks :)

/lgtm but it won't merge before another round of rebase + review!

**Review by kpouget — APPROVED** [review]
/lgtm

thanks @omertuc for this first PR, that's clean work, really appreciated!

**Pull request merged**
